### DISCUSSION TITLE
armcc编译器add_asflag()无法正常传入--predefine "IRQ_SAVE_INIT_PRIORITY SETA 0x20"给armasm.exe

### DISCUSSION THREAD

**Discussion by lucas-wu-wakakak** [xmake-io/xmake#2973] [Q&A] [answered]
### Xmake 版本

2.7.2

### 操作系统版本和架构

WIN10

### 描述问题

armcc编译器想要添加自定义预定义变量配置asm文件。编译器传入的标准语句为--predefine "IRQ_SAVE_INIT_PRIORITY SETA 0x20"
采用add_asflag()语句时，以下几种方式都无法正确传入命令：
add_asflag("--predefine IRQ_SAVE_INIT_PRIORITY SETA 0x20")
add_asflag([[--predefine "IRQ_SAVE_INIT_PRIORITY SETA 0x20"]])
add_asflag("--predefine [[IRQ_SAVE_INIT_PRIORITY SETA 0x20]]")

### 期待的结果

可以正常传入--predefine "IRQ_SAVE_INIT_PRIORITY SETA 0x20"

**Comment by waruqi** [comment]
add_defines("IRQ_SAVE_INIT_PRIORITY=0x20")

**Comment by lucas-wu-wakakak** [comment, marked as answer]
add_defines无法传入，反汇编相应的宏没有生效。

**Reply by waruqi** [reply]
贴下 xmake -v 的 完整编译命令 flags

**Comment by lucas-wu-wakakak** [comment]
传入的宏是-DIRQ_SAVE_INIT_PRIORITY=0x20 , 传入D宏是无法生效的.--predefine跟-D有点不一样, --predefine是全局生效的。
armcc手册：
--predefine is not equivalent  to the compiler option -Dname, --predefine defines a global variable whereas -Dname defines a macro the the C preprocessor expands

**Reply by lucas-wu-wakakak** [reply]
这个宏是要传给一个汇编文件的，-D只能作用于C/H文件

**Reply by waruqi** [reply]
默认配置，armcc 目前仅用于编译 c，所以传的 -D，asm 代码，目前应该走的 armasm.exe，会传 --predefine

为啥你这会用 armcc 编译 asm? 改了 --as ？凡事先贴 logs，和 xmake.lua ，

**Reply by lucas-wu-wakakak** [reply]
logs不方便贴，公司环境加密粘贴不出来，很多语句都是手敲的。我可能表述不严谨, 绝大部分都是用armcc编译c文件，里面有一两个asm文件需要单独编译，没有改环境，就是用的armasm.exe编译.  现在想把这个宏传给armasm.exe

**Reply by waruqi** [reply]
> asm文件需要单独编译，没有改环境，就是用的armasm.exe编译. 现在想把这个宏传给armasm.exe

如果是用的 armasm.exe 目前就是传的 --predefine 不是 -D 。。你自己看下。。

既然无法给 logs ，你只能自己仔细检查，反正不可能传 -D 给 armasm.exe 的，如果确实传了。自己调下相关 lua 代码，然后提 pr 过来。

https://github.com/xmake-io/xmake/blob/125f6017739c012357b69382972dc849841e32ce/xmake/modules/core/tools/armasm.lua#L102-L114

**Reply by lucas-wu-wakakak** [reply]
好的，我仔细确认下，确认完后来同步

**Comment by lucas-wu-wakakak** [comment]
已找到问题，汇编代码逻辑问题导致传入未生效。add_defines可以正常传入--pd 的定义，add_defines会在编译，汇编每个阶段都会传入一遍宏，同时符合各个阶段不用编译/链接器的定义要求。
